### PR TITLE
Define encodeInto() API

### LINKED ISSUE #69: TextEncoder#encode - write to existing Uint8Array
For binary protocols, such as msgpack, performance of writing/reading strings can be crucial, however those strings rarely (virtually never) exist on a binary buffer on their own. When decoding a string from a binary buffer of Uint8Array (obtained over, say, websockets) it's possible to cheaply create a subarray and pass it over to `TextDecoder#decode`, as such the decoding story is pretty straight forward.

The writing story is much more complicated.

In order to write a string into an existing buffer, one must take the result `Uint8Array` from `TextEncoder#encode` and `TypedArray#set` it at a desired offset into another buffer, which introduces unnecessary copying and, as result, is slower than plain JavaScript encoding implementations (at least for UTF-8).

A much better solution would be analog to the Node.js [`Buffer#write` implementation](https://nodejs.org/api/buffer.html#buffer_buf_write_string_offset_length_encoding) where the encoder can write a string to a buffer at a desired offset. Example signature could be something like:

```
TextEncoder#write(DOMString source, Uint8Array target [, Number offset [, Number length]]) -> Number bytesWritten
```

### PR DESCRIPTION
**Pull request by annevk**
This enables converting strings into byte sequences of pre-allocated buffers.

Also cleanup TextEncoder a bit.

Tests: ...

Fixes #69.

***
<a href="https://whatpr.org/encoding/166.html" title="Last updated on Jan 8, 2019, 2:04 PM UTC (7bb77bb)">Preview</a> | <a href="https://whatpr.org/encoding/166/c2879dd...7bb77bb.html" title="Last updated on Jan 8, 2019, 2:04 PM UTC (7bb77bb)">Diff</a>

### CODE REVIEW

**Comment by annevk** [comment]
The reason this does not contain the StringView dictionary is because it can be added later if there's indeed a need for it. This also does not require any particular destination length as either way out-of-band needs to be dealt with and it seems better to be flexible for simple string use cases that would otherwise have to write completely different code.

**Comment by hsivonen** [comment]
Looks correct to me, but it took me a moment to realize that the algorithm always loops over all tokens and the loop just does not necessarily write all the output. While an early return might be just an optimization, it might make the algorithm more comprehensible.

**Comment by annevk** [comment]
Thanks for pointing that out, I think it's actually wrong. Say we had 3 bytes left and the next input required 4 bytes and the input after that 1 byte. This would skip the 4 bytes input and then write the 1 byte one.

File: encoding.bs [hunk 1]
```diff
@@ -1319,18 +1323,14 @@ requires buffering of scalar values.
 
  <dt><code><var>encoder</var> . <a method for=TextEncoder lt=encode()>encode([<var>input</var> = ""])</a></code>
  <dd><p>Returns the result of running <a>UTF-8</a>'s <a for=/>encoder</a>.
+
+ <dt><code><var>encoder</var> . <a method=for=TextEncoder lt="encodeInto(source, destination)">encodeInto(<var>source</var>, <var>destination</var>)</a></code>
+ <dd><p>Runs the <a>UTF-8 encoder</a> on <var>source</var>, stores the result on
```
Comment (ricea) [inline encoding.bs]: "storing the result into destination" would sound more natural to me.

**Comment by annevk** [comment]
Do you know what the wording for those places should be? IDL doesn't really define any language around IDL ArrayBuffer objects... I guess I can file an issue upstream at least.

The other thing I realized is that I should add at least 1 example and document more clearly what read and written represent.

**Comment by inexorabletash** [comment]
I think this is implied by your last comment, but just to confirm: as written, _destination_'s view bounds (offset, length) onto its ArrayBuffer is seemingly ignored by the current algorithm. i.e. in this case:
```js
const ab = new ArrayBuffer(5);
const ua = new Uint8Array(ab, 1, 2);
new TextEncoder().encodeInto("ABCDE", ua);
```
... I would expect `ab` to end up as [0, 0x65, 0x66, 0, 0]. 

(I don't see test cases for this either)

**Comment by annevk** [comment]
Yeah, that clearly needs to be worded better. (I added tests.)

**Comment by annevk** [comment]
I think this is good now apart from the example which I'd like feedback on over at https://github.com/whatwg/encoding/issues/69#issuecomment-447275930. (And there's a Bikeshed issue I'll resolve with Tab.)

**Comment by annevk** [comment]
I haven't explained why there's no `decodeInto()`. I suppose "bytes in encoding X" -> "bytes in UTF-8" could make sense, but let's only add that if there's a request.

Anyone want to do a final review pass? (Tests need one too.)

File: encoding.bs [hunk 2]
```diff
@@ -1347,20 +1350,108 @@ must run these steps:
...
-    byte sequence, and then return a {{Uint8Array}} object wrapping an
-    {{ArrayBuffer}} containing <var>output</var>.
-    <!-- XXX https://www.w3.org/Bugs/Public/show_bug.cgi?id=26966 -->
+    <p>Otherwise:
 
-    <p class=note><a>UTF-8</a> cannot return <a>error</a>.
+    <ol>
+     <li>
+      <p>If <var>destinationBytes</var>'s <a for="byte sequence">length</a> &minus;
+      <var>written</var> is greater than the number of bytes in <var>result</var>, then:
```
Comment (ricea) [inline encoding.bs]: I think this needs to be "greater than or equal to".

Comment (annevk) [reply encoding.bs]: Yeah, indeed, thanks! (E.g., if we have space for 12 bytes, already written 6 bytes, then writing 6 more bytes should be fine whereas the current wording prohibits that.)

**Comment by annevk** [comment]
I'll land this and the tests tomorrow, unless there are further comments.

**Comment by annevk** [comment]
cc @whatwg/documentation

**Comment by annevk** [comment]
Implementation bugs:
* https://bugzilla.mozilla.org/show_bug.cgi?id=1514664
* https://bugs.webkit.org/show_bug.cgi?id=193274
* https://bugs.chromium.org/p/chromium/issues/detail?id=920107

**Pull request merged**